### PR TITLE
adds cloudbackup timers, go source and service files

### PR DESCRIPTION
**Pull request by paukul**
Testing this is possible via the cloudbackup channel.

In order for this to work, the aws / otc credentials have to be set up via skvs. I've created a bucket for testing purposes which can be used with the following credentials:

```
skvs_cli set cloudbackup/aws_bucket_name backup-staging
skvs_cli set cloudbackup/aws_secret_key UBWED9FCSGK9PJCZMK38SBNAWJW5JW0LM0Q4DQEJ
skvs_cli set cloudbackup/aws_access_key UALWRMPN8QE1ZLAE8BG2
```

In a nutshell, what the cloudbackup does is:

Outside the cloudbackup docker container:

1. take a zfs snapshot in the form of `volumename@hashofpgppubkey-date` every morning at 2 am
2. incrementally dump the snapshots + the tar'd content of /etc/protonet to a backup directory
3. delete old snapshots and keeps the 5 latest for every volume (cleaning up old ones happens after a new snapshot has successfully created and sent to disk)

Inside the cloudbackup docker container:

1. gzip the snapshot files
2. pgp encrypt the snapshot files
3. upload them in chunks of 20mb to the amazon open telecom platform
4. cleanup the backup directory

### CODE REVIEW

**Comment by paukul** [comment]
also, the part with the docker container for the actual backup server is obviously still missing

update: no longer missing ;-)

**Comment by colszowka** [comment]
No real review, just wondering, shouldn't this have some unit tests, too? :angel:

**Comment by tiff** [comment]
what's the status here @hinnerk?

**Comment by paukul** [comment]
I've been sick yesterday and today. That's why this PR is still not finished

**Comment by hinnerk** [comment]
I already answered offline, ftr: for now all seems to be going according to plan.

**Comment by paukul** [comment]
I've updated the description of the PR so it should be reviewable

**Comment by colszowka** [comment]
Maybe the cloud-backup.sh logic could use some error handling in case things go wrong? Not sure how to tackle this UI-wise, but I'd imagine users would like to get some form of alert if their backup is not working anymore before actually needing to restore it :)

As to the go-code, looks good to my naive eyes

**Comment by paukul** [comment]
@colszowka 
The PR for Soul includes the user facing parts. Those have an indicator when the last backup completed successfully if the last backup failed

https://github.com/protonet/german-shepherd/pull/1165

**Comment by colszowka** [comment]
@paukul Ah, excellent - thanks :)

File: services/cloudbackup.service [hunk 1]
```diff
@@ -19,6 +19,7 @@ ExecStartPre=/usr/bin/env bash -c "/usr/bin/docker run -d \
       --env \"AWS_ACCESS_KEY=$(/opt/bin/skvs_cli get cloudbackup/aws_access_key)\" \
       --env \"AWS_SECRET_KEY=$(/opt/bin/skvs_cli get cloudbackup/aws_secret_key)\" \
       --env \"AWS_BUCKET_NAME=$(/opt/bin/skvs_cli get cloudbackup/aws_bucket_name)\" \
+      --env AWS_REGION=eu-de \
```
Comment (kdomanski) [inline services/cloudbackup.service:22]: Is that a Telekom Cloud region?

Comment (paukul) [reply services/cloudbackup.service:22]: yeh, it's "THE" otc region. They only have one afaik :-)

File: services/init-protonet.service [hunk 2]
```diff
@@ -27,7 +27,7 @@ ExecStartPre=/bin/mkdir -p /data/soul/public
 ExecStartPre=/bin/mkdir -p /data/soul/uploads
 ExecStartPre=/bin/mkdir -p /data/soul/protosync
 ExecStartPre=/bin/mkdir -p /data/pgp
-ExecStartPre=/bin/bash -c '[[ ! -d /backup ]] && zfs create -p -o mountpoint=/backup protonet_storage/backup'
+ExecStartPre=-/bin/bash -c '[[ ! -d /backup ]] && zfs create -p -o mountpoint=/backup protonet_storage/backup'
```
Comment (kdomanski) [inline services/init-protonet.service:30]: It seems like `zfs create` does nothing if a filesystem already exists and it exits with 0, so `/usr/sbin/zfs create -p -o mountpoint=/backup protonet_storage/backup` would be sufficient.

Comment (paukul) [reply services/init-protonet.service:30]: I had this one failing on the AMI tho. that's why I changed it. I'll double check if I mixed something up here

**Comment by hinnerk** [comment]
Merging, just because it's a forked branch and we need to modify stuff in it.

File: scripts/create-backup.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+set -e
+
+keyhash=$(md5sum /data/pgp/public.key)
```
Comment (hinnerk) [inline scripts/create-backup.sh:5]: `/data/cloudbackup/…`please.

File: services/init-protonet.service [hunk 4]
```diff
@@ -34,6 +36,7 @@ ExecStartPre=/bin/chown 1000:1000 /data/soul/log
 ExecStartPre=/bin/chown 1000:1000 /data/soul/public
 ExecStartPre=/bin/chown 1000:1000 /data/soul/uploads
 ExecStartPre=/bin/chown 1000:1000 /data/soul/protosync
+ExecStartPre=/bin/chown 1000:1000 /data/pgp
```
Comment (hinnerk) [inline services/init-protonet.service:39]: `/data/cloudbackup/…`please.

File: services/run-cloudbackup.timer [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Starts creating a zfs snapshot for the cloud backup
+ConditionPathExists=/etc/protonet/soul/enabled
```
Comment (hinnerk) [inline services/run-cloudbackup.timer:3]: `/etc/protonet/cloudbackup/enabled`

File: services/run-cloudbackup.timer [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+[Timer]
+OnCalendar=*-*-* 02:00:00
+Unit=run-cloudbackup.service
```
Comment (hinnerk) [inline services/run-cloudbackup.timer:7]: `cloudbackup-run` in order to support a sorted list.

File: services/run-cloudbackup.service [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Creates a zfs snapshot and store it in the backup inbox directory
+ConditionPathExists=/data/pgp/public.key
```
Comment (hinnerk) [inline services/run-cloudbackup.service:3]: - `/etc/protonet/cloudbackup/enabled`
- `/data/cloudbackup/…`

File: services/run-cloudbackup.timer (hunk 5 above)
Comment (paukul) [reply services/run-cloudbackup.timer:3]: alright. I figured soul/enabled was the general trigger for soul and everything connected. I'll change it as suggested. do i need to enable this somewhere then by default or will you take care. and if I have to, could you point me into the general direction?

**Comment by hinnerk** [comment]
All fixed in https://github.com/experimental-platform/platform-configure/pull/290

File: services/run-cloudbackup.timer (hunk 5 above)
Comment (hinnerk) [reply services/run-cloudbackup.timer:3]: Ups, sorry, all done. https://github.com/experimental-platform/platform-configure/pull/290

**Pull request merged**